File: Source/Orleankka.Tests/Scenarios/Reentrant_messages.cs
using System;
using System.Linq;
using System.Runtime.Remoting.Messaging;

using NUnit.Framework;

namespace Orleankka.Scenarios
{
    using Meta;
    using Testing;

    [RequiresSilo]
    public class Reentrant_messages : ActorSystemScenario
    {
        [Test]
        public async void Reentrant_could_be_defined_via_attribute()
        {
            var actor = system.FreshActorOf<TestReentrantDefinedViaAttributeActor>();
            Assert.That(await actor.Ask(new RegularMessage()), Is.False);
            Assert.That(await actor.Ask(new ReentrantMessage()), Is.True);
        }
        
        [Test]
        public async void Reentrant_could_be_defined_via_prototype()
        {
            var actor = system.FreshActorOf<TestReentrantDefinedViaPrototypeActor>();
            Assert.That(await actor.Ask(new RegularMessage()), Is.False);
            Assert.That(await actor.Ask(new ReentrantMessage()), Is.True);
        }

        [Test]
        public void Reentrant_cannot_be_defined_outside_of_prototype_definition()
        {
            var actor = system.FreshActorOf<TestReentrantDefinedOutsideOfPrototypeActor>();
            Assert.Throws<InvalidOperationException>(async ()=> await actor.Tell("boo"));
        }

        class RegularMessage : Query<bool>
        {}

        class ReentrantMessage : Query<bool>
        {}

        abstract class TestReentrantActorBase : Actor
        {
            protected internal override void Define()
            {
                On((RegularMessage x)   => ReceivedReentrant(x));
                On((ReentrantMessage x) => ReceivedReentrant(x));
            }

            static bool ReceivedReentrant(object message)
            {
                return CallContext.LogicalGetData("LastMessageReceivedReentrant") == message;
            }
        }

        [Reentrant(typeof(ReentrantMessage))]
        class TestReentrantDefinedViaAttributeActor : TestReentrantActorBase
        {}

        class TestReentrantDefinedViaPrototypeActor : TestReentrantActorBase
        {
            protected internal override void Define()
            {
                Reentrant(x => x is ReentrantMessage);
                
                base.Define();
            }
        }

        class TestReentrantDefinedOutsideOfPrototypeActor : TestReentrantActorBase
        {
            public void Handle(string message)
            {
                Reentrant(x => x is ReentrantMessage);
            }
        }
    }
}

File: Source/FSharp.Api.Tests/ReentrantMessages.fs
module Orleankka.FSharp.ReentrantMessages

open NUnit.Framework
open System
open Orleankka
open Orleankka.FSharp.Tests.Infrastructure
open System.Runtime.Remoting.Messaging

type RegularMessage = {
   Data: string 
} 

type ReentrantMessage = {
   Data: string
}

type ReentrantMessage2() = class end

[<Reentrant(typeof<ReentrantMessage>)>]
[<Reentrant(typeof<ReentrantMessage2>)>]  
type TestActor() = 
   inherit Actor<obj>()

   let receivedReentrant msg = CallContext.LogicalGetData("LastMessageReceivedReentrant") = msg

   override this.Receive message reply = task {
      match message with
      | :? RegularMessage as m    -> m |> receivedReentrant |> reply
      
      | :? ReentrantMessage as m  -> m |> receivedReentrant |> reply

      | :? ReentrantMessage2 as m -> m |> receivedReentrant |> reply
      
      | _                         -> failwith "Received unexpected message type"
   }

[<TestFixture>]
[<RequiresSilo>]
type Tests() =   
   [<DefaultValue>] val mutable system: IActorSystem

   [<SetUp>]
   member this.SetUp() = 
      this.system <- TestActorSystem.instance

   [<Test>]
   member this.``Could be defined via attribute.``() = 
      let actor = this.system.ActorOf<TestActor>("test")
      
      match Task.run(fun _ -> task {return! actor <? { RegularMessage.Data = "test" }}) with
      | Choice1Of2 (result:bool) -> Assert.That(result, Is.False)
      | Choice2Of2 ex            -> Assert.Fail(ex.ToString())

      match Task.run(fun _ -> task {return! actor <? { ReentrantMessage.Data = "test" }}) with
      | Choice1Of2 (result:bool) -> Assert.That(result, Is.True)
      | Choice2Of2 ex            -> Assert.Fail(ex.ToString())

      match Task.run(fun _ -> task {return! actor <? ReentrantMessage2() }) with
      | Choice1Of2 (result:bool) -> Assert.That(result, Is.True)
      | Choice2Of2 ex            -> Assert.Fail(ex.ToString())